File: src/core/SemaSession.ts
import { SemaEngine } from './SemaEngine';
import { CreateSessionOptions } from '../types/session';
import { ForkOptions, ForkPreview, ForkResult } from '../types/fork';
import { SessionEventBus, createSessionEventBus } from '../events/EventSystem';
import { generateSessionId } from '../util/session';
import { getStateManager } from '../manager/StateManager';
import { getCheckpointManager } from '../manager/CheckpointManager';
import { getTaskManager } from '../manager/TaskManager';
import { ToolPermissionResponse, PickOptionResponseData, PlanExitResponseData } from '../events/types';
import { TaskListItem } from '../types/task';
import type { AgentMode, PermissionLevel } from '../types';
import type { InputImageAttachment } from '../types/message';
import { logInfo } from '../util/log';

/**
 * 会话级 API
 * 由 SemaCore.createSession() 创建并返回，承载单个会话的全部交互能力。
 * 进程级能力（模型/MCP/插件/agents 配置）仍在 SemaCore 上。
 */
export class SemaSession {
  constructor(
    public readonly sessionId: string,
    private readonly engine: SemaEngine,
    private readonly bus: SessionEventBus,
  ) {}

  /**
   * 创建并初始化一个会话：装配事件总线、引擎与会话对象。
   * 初始化失败时自动清理并向上抛出异常，由调用方决定如何处理。
   */
  static async create(opts: CreateSessionOptions = {}): Promise<SemaSession> {
    const sessionId = opts.sessionId || generateSessionId();
    const bus = createSessionEventBus(sessionId);
    const engine = new SemaEngine(sessionId, bus);
    const session = new SemaSession(sessionId, engine, bus);

    try {
      await engine.createSession({ sessionId: opts.sessionId, agentMode: opts.agentMode });
    } catch (error) {
      session.dispose();
      throw error;
    }

    return session;
  }

  // ==================== 事件接口（仅本会话）====================
  on = <T>(event: string, listener: (data: T) => void) => (this.bus.on(event, listener), this);
  once = <T>(event: string, listener: (data: T) => void) => (this.bus.once(event, listener), this);
  off = <T>(event: string, listener: (data: T) => void) => (this.bus.off(event, listener), this);

  // 权限/交互响应接口
  respondToToolPermission = (response: ToolPermissionResponse) =>
    this.bus.emit('tool:permission:response', response);
  respondToPickOption = (response: PickOptionResponseData) =>
    this.bus.emit('pick:option:response', response);
  respondToPlanExit = (response: PlanExitResponseData) =>
    this.bus.emit('plan:exit:response', response);

  // ==================== 会话交互 ====================
  processUserInput = (input: string, originalInput?: string, attachments?: InputImageAttachment[]): void =>
    this.engine.processUserInput(input, originalInput, false, attachments);

  interrupt = (): void => this.engine.interruptSession();

  // ==================== 会话级配置 ====================
  updateAgentMode = (mode: AgentMode): void => this.engine.updateAgentMode(mode);
  updatePermissionLevel = (level: PermissionLevel): void => this.engine.updatePermissionLevel(level);

  // ==================== 会话 Fork / 撤销 ====================
  /** 预览：在该用户消息处恢复文件会改动哪些文件、各自增删行数（只读，无副作用） */
  getForkPreview = (messageUuid: string): ForkPreview =>
    getCheckpointManager().preview(this.sessionId, messageUuid);

  /**
   * 原地回退（Fork / 撤销）：把会话历史截断到该用户消息之前；
   * restoreFiles=true 时同时回滚文件。会话 id 不变，继续使用。
   */
  fork = (messageUuid: string, options?: ForkOptions): Promise<ForkResult> =>
    this.engine.rewind(messageUuid, options);

  // ==================== 后台任务（仅本会话）====================
  getTaskList = (): TaskListItem[] => getTaskManager().getTaskList(this.sessionId);
  watchTask = (taskId: string, onDelta: (delta: string) => void): () => void =>
    getTaskManager().watchTask(taskId, onDelta);
  stopTask = (taskId: string): boolean => getTaskManager().stopTask(taskId);
  stopAllTasks = (): number => getTaskManager().stopAllTasks(this.sessionId);
  transferAgentToBackground = (taskId: string): boolean => getTaskManager().transferToBackground(taskId);
  transferAllForegroundAgents = (): string[] => getTaskManager().transferAllForeground(this.sessionId);

  /**
   * 关闭会话（会话级 dispose，不触碰全局单例）
   */
  dispose = (): void => {
    logInfo(`关闭会话: ${this.sessionId}`);
    this.engine.dispose();
    getTaskManager().disposeSession(this.sessionId);
    getStateManager().removeSession(this.sessionId);
    getCheckpointManager().forgetSession(this.sessionId);
    this.bus.dispose();
  };
}

File: sdks/java/src/main/java/semacore/SemaSession.java
package semacore;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.reflect.TypeToken;
import semacore.protocol.Registration;
import semacore.protocol.SemaBridgeClient;
import semacore.type.*;
import semacore.event.ToolPermissionResponse;
import semacore.event.PickOptionResponseData;
import semacore.event.PlanExitResponseData;

import java.lang.reflect.Type;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.function.Predicate;

/**
 * sema-core 会话的镜像 API：用法与 Node 侧 {@code const session = await core.createSession()} 一致。
 *
 * <p>方法名/参数名 = core 的 SemaSession；返回值 = core 返回类型（{@code semacore.type} 强类型 DTO）；
 * 事件名 = core 原始事件名，{@link #on} 自动按本会话 sessionId 过滤，回调 data 为 {@link JsonElement}。
 * 同步/阻塞语义与死锁注意事项见 {@link SemaCore}。</p>
 *
 * <p><b>创建期事件语义</b>：会话对象构造时即捕获 {@code session:ready} 并缓存，之后（含晚）订阅立即重放，
 * 保证 Node 式「createSession 后再订阅」不丢事件。</p>
 */
public final class SemaSession {

    private static final Gson GSON = new Gson();
    private static final Type LIST_STRING = new TypeToken<List<String>>() {}.getType();

    private final SemaBridgeClient client;
    private final String sessionId;
    private final CompletableFuture<JsonElement> ready = new CompletableFuture<>();
    private final List<Sub> subs = new ArrayList<>();

    private record Sub(String event, Consumer<JsonElement> handler, Registration reg) {}

    SemaSession(SemaBridgeClient client, String sessionId) {
        this.client = client;
        this.sessionId = sessionId;
        // 构造发生在 createSession ack 的分发链上，早于桥推送 session:ready（同一 gRPC 流串行分发），故此处订阅必不丢失。
        client.once("session:ready", sessionId, (data, sid) -> ready.complete(Json.parse(data)));
    }

    public String sessionId() {
        return sessionId;
    }

    // ── 会话交互 ─────────────────────────────────────────────────────────

    public void processUserInput(String input) {
        processUserInput(input, null, null);
    }

    public void processUserInput(String input, String originalInput) {
        processUserInput(input, originalInput, null);
    }

    public void processUserInput(String input, String originalInput, List<InputImageAttachment> attachments) {
        send("processUserInput", Json.obj("content", input, "orgContent", originalInput, "attachments", attachments));
    }

    public void interrupt() {
        send("interrupt", null);
    }

    // ── 交互应答 ─────────────────────────────────────────────────────────

    public void respondToToolPermission(ToolPermissionResponse response) {
        send("respondToToolPermission", response);
    }

    public void respondToPickOption(PickOptionResponseData response) {
        send("respondToPickOption", response);
    }

    public void respondToPlanExit(PlanExitResponseData response) {
        send("respondToPlanExit", response);
    }

    // ── 模式 / 权限 ──────────────────────────────────────────────────────

    public void updateAgentMode(AgentMode mode) {
        send("updateAgentMode", Json.obj("mode", mode));
    }

    public void updatePermissionLevel(PermissionLevel level) {
        send("updatePermissionLevel", Json.obj("level", level));
    }

    // ── fork / 撤销回退 ──────────────────────────────────────────────────

    public ForkPreview getForkPreview(String messageUuid) {
        return call("getForkPreview", Json.obj("messageUuid", messageUuid), ForkPreview.class);
    }

    public ForkResult fork(String messageUuid) {
        return fork(messageUuid, null);
    }

    public ForkResult fork(String messageUuid, ForkOptions options) {
        JsonElement el = req("fork", Json.obj("messageUuid", messageUuid, "options", options));
        // 判别联合按 ok 分派到具体变体（≙ core ForkResult）
        return SemaJson.bool(el, "ok", false)
                ? GSON.fromJson(el, ForkResultOk.class)
                : GSON.fromJson(el, ForkResultErr.class);
    }

    // ── 后台任务（会话级）────────────────────────────────────────────────

    public List<TaskListItem> getTaskList() {
        return callList("getTaskList", null, TaskListItem.class);
    }

    /** 停掉本会话全部后台子 agent；返回停掉数量。 */
    public int stopAllTasks() {
        return SemaJson.i32(req("stopAllTasks", null), "count", 0);
    }

    public boolean stopTask(String taskId) {
        return SemaJson.bool(req("stopTask", Json.obj("taskId", taskId)), "ok", false);
    }

    public boolean transferAgentToBackground(String taskId) {
        return SemaJson.bool(req("transferAgentToBackground", Json.obj("taskId", taskId)), "ok", false);
    }

    public List<String> transferAllForegroundAgents() {
        return GSON.fromJson(SemaJson.get(req("transferAllForegroundAgents", null), "taskIds"), LIST_STRING);
    }

    /** 开始流式观察任务输出；关闭返回的 Registration 会停止观察并退订事件。 */
    public Registration watchTask(String taskId, Consumer<String> onDelta) {
        Objects.requireNonNull(taskId);
        Objects.requireNonNull(onDelta);
        AtomicBoolean active = new AtomicBoolean(true);
        Registration registration = client.on("task:watch:delta", sessionId, (data, sid) -> {
            JsonElement element = Json.parse(data);
            if (!element.isJsonObject()) return;
            JsonElement eventTaskId = element.getAsJsonObject().get("taskId");
            if (eventTaskId == null || !taskId.equals(eventTaskId.getAsString())) return;
            JsonElement delta = element.getAsJsonObject().get("delta");
            if (delta != null && !delta.isJsonNull()) onDelta.accept(delta.getAsString());
        });
        send("watchTask", Json.obj("taskId", taskId));
        return () -> {
            if (!active.compareAndSet(true, false)) return;
            registration.unregister();
            send("unwatchTask", Json.obj("taskId", taskId));
        };
    }

    // ── 事件（core 原始事件名，自动按本会话过滤）─────────────────────────

    public Registration on(String event, Consumer<JsonElement> handler) {
        if ("session:ready".equals(event)) return subscribeReady(handler);
        Registration reg = client.on(event, sessionId, (data, sid) -> handler.accept(Json.parse(data)));
        subs.add(new Sub(event, handler, reg));
        return reg;
    }

    public Registration once(String event, Consumer<JsonElement> handler) {
        if ("session:ready".equals(event)) return subscribeReady(handler);
        Registration reg = client.once(event, sessionId, (data, sid) -> handler.accept(Json.parse(data)));
        subs.add(new Sub(event, handler, reg));
        return reg;
    }

    /** 按 (event, handler) 取消订阅（≙ Node off）；移除首个匹配的 on/once 注册。 */
    public void off(String event, Consumer<JsonElement> handler) {
        for (int i = 0; i < subs.size(); i++) {
            Sub s = subs.get(i);
            if (s.event().equals(event) && s.handler() == handler) {
                s.reg().unregister();
                subs.remove(i);
                return;
            }
        }
    }

    /** 等待本会话某事件触发一次；timeout 为 null 表示不超时。 */
    public JsonElement waitFor(String event, Duration timeout) {
        return waitFor(event, timeout, null);
    }

    /** 等待本会话某事件（可选 predicate 对解析后的 data 过滤）。 */
    public JsonElement waitFor(String event, Duration timeout, Predicate<JsonElement> predicate) {
        if ("session:ready".equals(event)) {
            JsonElement cached = ready.getNow(null);
            if (cached != null && (predicate == null || predicate.test(cached))) return cached;
        }
        CompletableFuture<JsonElement> future = new CompletableFuture<>();
        Registration reg = client.on(event, sessionId, (data, sid) -> {
            JsonElement d = Json.parse(data);
            if (predicate == null || predicate.test(d)) future.complete(d);
        });
        future.whenComplete((r, e) -> reg.unregister());
        if (timeout != null) future.orTimeout(timeout.toMillis(), TimeUnit.MILLISECONDS);
        try {
            return future.join();
        } catch (CompletionException e) {
            Throwable c = e.getCause();
            if (c instanceof RuntimeException re) throw re;
            throw e;
        }
    }

    /** session:ready 走缓存重放（本质一次性事件，on/once 等价）。 */
    private Registration subscribeReady(Consumer<JsonElement> handler) {
        AtomicBoolean active = new AtomicBoolean(true);
        ready.thenAccept(data -> {
            if (active.get()) handler.accept(data);
        });
        return () -> active.set(false);
    }

    // ── 生命周期 ─────────────────────────────────────────────────────────

    /** 关闭本会话（≙ core.closeSession；只关会话不关连接）。 */
    public boolean close() {
        JsonElement ack = Json.parse(SemaCore.join(client.call("closeSession", null, sessionId)));
        return SemaJson.bool(ack, "ok", false);
    }

    // ── 内部 ─────────────────────────────────────────────────────────────

    private JsonElement req(String action, Object payload) {
        return Json.parse(SemaCore.join(client.call(action, Json.stringify(payload), sessionId)));
    }

    private <T> T call(String action, Object payload, Class<T> type) {
        return GSON.fromJson(req(action, payload), type);
    }

    private <T> List<T> callList(String action, Object payload, Class<T> elem) {
        Type t = TypeToken.getParameterized(List.class, elem).getType();
        List<T> list = GSON.fromJson(req(action, payload), t);
        return list != null ? list : new ArrayList<>();
    }

    /** fire-and-forget（Node void 语义）：发送不阻塞，不等 ack。 */
    private void send(String action, Object payload) {
        client.call(action, Json.stringify(payload), sessionId).exceptionally(e -> null);
    }
}
